### ISSUE TITLE
Segmentation fault when scaling and strict sandbox is used

### ISSUE THREAD

**Issue by sebastinas** [pwmt/zathura#162]
*On GitLab by @valoq on Nov 13, 2019, 24:24*

---

Using sandbox=strict and scaling results in the following error when the mouse is hovered over zathuras window:

```
(org.pwmt.zathura:4878): Gdk-WARNING **: 00:10:17.210: Failed to load cursor theme Adwaita with scale 2
Segmentation fault (core dumped)
```

When looking at this with strace, there is an attempt to open a file for writing:

` openat(AT_FDCWD, 0xbd1ec4e8630, O_RDWR|O_CREAT|O_EXCL|O_CLOEXEC, 0600) = -1 EACCES (Permission denied) `

(I have yet to find out which file it is.)

This is however not permitted by the seccomp filter.

Does anyone have some idea on how to find out whats causing this?
The issue could be resolved by allowing openat generally but that would significantly decrease the effect of seccomp.

I would also like to find out what exactly is trying to write something just because scaling is used. This smells like a bug in GTK (or girara?) and I would rather avoid this altogether then to allow writing files.

There is a hint in the journal pointing to wayland as well:

`zathura[8606]: segfault at 0 ip 00006ea451ecd69a sp 000075565d1c3710 error 4 in libwayland-cursor.so.0.0.0[6ea451ecd000+2000]`

Any hints or ideas are very much welcome.

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Nov 13, 2019, 24:56*

---

Just found the following output right before the crash

`debug: ../zathura/zathura.c:192: zathura_update_view_ppi(): on Wayland, correcting PPI for device scale factor = 2`

which is triggered in:

```
  if (GDK_IS_WAYLAND_DISPLAY(display))
   {
     /* not using the cached value for the scale factor here to avoid issues
      * if this function is called before the cached value is updated */
     const int device_factor = gtk_widget_get_scale_factor(zathura->ui.session->gtk.view);
     girara_debug("on Wayland, correcting PPI for device scale factor = %d", device_factor);
     if (device_factor != 0) {
       ppi /= device_factor;
     }
   }
```

I will have to take a closer look at it later, but it looks like girara could be the issue here. In any case it should not crash because of a file or permissions missing.
For some reason the bug also does not trigger when running zathura as root and there is also no mention of seccomp in audit, even though the bug also disappears when seccomp allows openat. I suppose because the syscall is gallowed (for reading) the process is not killed but rather a permission denied is the result. But this should also be the case when running as root...

**Comment by sebastinas** [comment]
*On GitLab by @valoq on Nov 17, 2019, 16:33*

---

I don't see what's causing this, the scaling seems to be calculated correctly

Although unrelated, I think this 
```
if (device_factor != 0) {
  ppi /= device_factor;
}
```
is not working as expected, since the device factor will never be 0 (default is 1) and so this code is always executed even if dividing by 1 does not do anything.

Could it be that the GDK bug that this code is written for has changed and is causing this issue?
I do not have any experience in GDK/girara, maybe someone who does can see whats going wrong here.
The core issue is probably unrelated to seccomp, although it does not seem to cause a segfault without seccomp enabled.

**Comment by sebastinas** [comment]
I don't think that this is related. It looks to me like the seccomp filter currently disallows GTK+ to open files it requires to work correctly.

**Comment by sebastinas** [comment]
*On GitLab by @valoq on May 30, 2020, 13:09*

---

This issue was resolved by wayland in version 1.18

https://gitlab.freedesktop.org/wayland/wayland/-/issues/123

**Comment by sebastinas** [comment]
*On GitLab by @valoq on May 30, 2020, 13:09*

---

closed
